### PR TITLE
Added classifiers to emit events when matching regular expressions.

### PR DESCRIPTION
**Pull request by rayzorben**

### CODE REVIEW

File: src/main/entities/DocumentData.js [hunk 1]
```diff
@@ -200,7 +211,8 @@ export default class DocumentData {
 
         // https://www.notthewizard.com/2014/06/17/are-files-appends-really-atomic/
         if (content.length >= 1024 && !release) {
-          release = await lockfile.lock(this.filename, lockOptions);
+          console.log('locking file for saving');
```
Comment (onefuncman) [inline src/main/entities/DocumentData.js:214]: no need to keep this ?

File: src/main/entities/PersistableEntity.js [hunk 2]
```diff
@@ -109,27 +107,66 @@ export default class PersistableEntity {
...
-      this._datastores[file] = database;
+    // If initialization is in progress, wait for it
+    if (this._initializationPromises[file]) {
+      return await this._initializationPromises[file];
     }
 
-    await release();
+    // Create initialization promise for this file
+    this._initializationPromises[file] = (async () => {
+      console.log('creating initialization promise');
```
Comment (onefuncman) [inline src/main/entities/PersistableEntity.js:127]: -log

File: src/main/entities/PersistableEntity.js [hunk 3]
```diff
@@ -109,27 +107,66 @@ export default class PersistableEntity {
...
+        await fs.promises.mkdir(path.dirname(file), { recursive: true });
+
+        try {
+          await fs.promises.access(file);
+        } catch (err) {
+          await fs.promises.writeFile(file, '');
+        }
+
+        // lock the file
+        console.log('locking file for database');
```
Comment (onefuncman) [inline src/main/entities/PersistableEntity.js:139]: -log

File: src/main/entities/DocumentData.js (hunk 1 above)
Comment (rayzorben) [reply src/main/entities/DocumentData.js:214]: oops I literally was in the process of removing those and then my ADHD kicked in and I forgot. Let me clean them up.

File: src/main/entities/PersistableEntity.js [hunk 4]
```diff
@@ -109,27 +107,66 @@ export default class PersistableEntity {
...
+          const database = await doc.load(false); // we control the lock
+          this._datastores[file] = database;
+        }
+
+        await release();
+        return this._datastores[file];
+      } finally {
+        // Schedule cleanup for the next tick after promise resolves
+        Promise.resolve().then(() => {
+          console.log('deleting initialization promise');
```
Comment (onefuncman) [inline src/main/entities/PersistableEntity.js:162]: -log

File: src/main/entities/PersistableEntity.js [hunk 5]
```diff
@@ -109,27 +107,66 @@ export default class PersistableEntity {
...
+          delete this._initializationPromises[file];
+        });
+      }
+    })();
 
-    return this._datastores[file];
+    console.log('returning initialization promise');
```
Comment (onefuncman) [inline src/main/entities/PersistableEntity.js:168]: -log

File: src/main/entities/PersistableEntity.js [hunk 6]
```diff
@@ -226,6 +263,7 @@ export default class PersistableEntity {
 
   async save() {
     if (this.dirty) {
+      console.log('SAVING:', this._document);
```
Comment (onefuncman) [inline src/main/entities/PersistableEntity.js:266]: -log

**Review by onefuncman — APPROVED** [review]

**Pull request merged**